### DISCUSSION TITLE
How to solve a forced ODE with measured forcing?

### DISCUSSION THREAD

**Discussion by aleksejs-fomins** [pnnl/neuromancer#166] [Q&A] [answered]
I would like to perform inference on a basic integrator $\dot{X} = k(F(t) - X(t))$, where $k$ is an unknown constant and $F(t)$ is a measured dataset - its sampled uniformly between $t_{start}$ and $t_{end}$. I have had a look at the [ODE parameter estimation tutorial](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/ODEs/Part_2_param_estim_ODE.ipynb), but I can't figure out how to modify the existing solution to include forcing, because minibatches do not seem to explicitly carry time information in that example, so I don't know what time interval each batch is related to in order to correctly construct the right hand side of the ODE. Could you please suggest how I can do that, or direct me to a piece of code that already does something like that?

Also, perhaps somebody could help me better understand this example. I find that the explanations are a bit too brief, especially with regarding to the variable 'xn'. By reading the code, I have made the following hypothesis

* 'X' is a tensor of the true values of the dynamical system variables, of shape [number_trajectories, number_timestep_in_batch, number_dimensions]
* 'xn' is the tensor of the predictions for X, obtained via integration. For each batch, 'xn' is given an initial value of the true X at the time of the start of the batch, and then forwards-integrated (e.g. using RK45)

**Comment by drgona** [comment, marked as answer]
Hi @aleksejs-fomins.
We have a few existing notebooks that include forcing (control inputs) for neural time steppers.

- [Neural ODE example](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/domain_examples/NODE_swing_equation.ipynb)
- [Neural state space model example](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/ODEs/Part_5_nonauto_NSSM.ipynb)
- [Deep Koopman model example](https://colab.research.google.com/github/pnnl/neuromancer/blob/master/examples/ODEs/Part_8_nonauto_DeepKoopman.ipynb)

Each of them features a different neural time stepper architecture, but the API is the same for all of them.
In case you have regularly sampled the forcing signal, you can just include it in the training dataset and label it with a variable string of your choice. We use 'U' for denoting inputs and 'X' for denoting states. This is a convention from control theory.
Your observations on 'X' and 'xn' are correct in the examples above. Please note you can label them as you wish and don't have to follow the convention.

In all of these examples we assume 3D tensors in the training data, the same of the tensors corresponds to [batch_size, nsteps, variable_size]. Where batch_size determines the number of trajectories, nsteps defines the length of the rollout horizon, and variable_size defines the number of variables (states, inputs).
Our variable class acts as a pointer to the data or model outputs and is intended to be used for defining symbolic constraints and objective function terms. The variable class supports all pytorch callables and automatically constructs symbolic computational graph. But because we know the tensor dimensions in this example, we can also slice the symbolic variables and index them, for instance, to impose terminal constraints one step loss terms - see examples for more details.

Hope this helps.
